long-ride-alerts/src/main/java/org/apache/flink/training/exercises/longrides/LongRidesExercise.java

```diff
@@ -98,17 +99,66 @@ public static void main(String[] args) throws Exception {
     @VisibleForTesting
     public static class AlertFunction extends KeyedProcessFunction<Long, TaxiRide, Long> {
 
+        private ValueState<TaxiRide> rideStatus;
+
         @Override
-        public void open(Configuration config) throws Exception {
-            throw new MissingSolutionException();
+        public void open(Configuration config)  {
+            ValueStateDescriptor<TaxiRide> rideStateDescriptor =
+                    new ValueStateDescriptor<>("ride event", TaxiRide.class);
+            rideStatus = getRuntimeContext().getState(rideStateDescriptor);
         }
 
         @Override
-        public void processElement(TaxiRide ride, Context context, Collector<Long> out)
-                throws Exception {}
+        public void processElement(TaxiRide ride, Context context, Collector<Long> out) throws Exception {
+            TaxiRide firstRideEvent = rideStatus.value();
+
+            if (firstRideEvent == null) {
+                rideStatus.update(ride);
+
+                if (ride.isStart) {
+                    // we will use this timer to check for rides that have gone on too long and may
+                    // not yet have an END event (or the END event could be missing)
+                    context.timerService().registerEventTimeTimer(checkTimerTime(ride));
+                }
+            } else {
+                if (ride.isStart) {
+                    if (rideIsLong(ride, firstRideEvent)) {
+                        out.collect(ride.rideId);
+                    }
+                } else {
+                    context.timerService().deleteEventTimeTimer(checkTimerTime(firstRideEvent));
+
+                    if (rideIsLong(firstRideEvent, ride)) {
+                        out.collect(ride.rideId);
+                    }
+                }
+
+                // both events have now been seen, we can clear the state
+                // this solution can leak state if an event is missing
+                // see DISCUSSION.md for more information
+                rideStatus.clear();
+            }
 
+        }
         @Override
         public void onTimer(long timestamp, OnTimerContext context, Collector<Long> out)
-                throws Exception {}
+                throws Exception {
+            out.collect(rideStatus.value().rideId);
+            rideStatus.clear();
+        }
+
+        private boolean rideIsLong(TaxiRide startEvent, TaxiRide endEvent) {
+            return Duration.between(startEvent.eventTime, endEvent.eventTime)
+                    .compareTo(Duration.ofHours(2))
+                    > 0;
+        }
+
+        private long checkTimerTime(TaxiRide ride) throws RuntimeException {
```

Comment:
btw there is no reason to have: `throws RuntimeException` it's unchecked exception.